OpenRA.Mods.Common/Traits/Player/TimeLimitManager.cs

```diff
@@ -0,0 +1,165 @@
+		public readonly bool SkipTimeRemainingNotifications = false;
+
+		[Desc("Will prevent showing/playing the built-in timer expired notification when set to true.")]
+		public readonly bool SkipTimerExpiredNotification = false;
+
+		IEnumerable<LobbyOption> ILobbyOptions.LobbyOptions(Ruleset rules)
+		{
+			var timelimits = TimeLimitOptions.ToDictionary(c => c.ToString(), c =>
+			{
+				if (c == 0)
+					return "No limit";
+				else
+					return c.ToString() + " minute{0}".F(c > 1 ? "s" : null);
+			});
+
+			yield return new LobbyOption("timelimit", TimeLimitLabel, TimeLimitDescription, true, TimeLimitDisplayOrder,
+				new ReadOnlyDictionary<string, string>(timelimits), TimeLimitDefault.ToString(), TimeLimitLocked);
+		}
+
+		public object Create(ActorInitializer init) { return new TimeLimitManager(init.Self, this); }
+	}
+
+	public class TimeLimitManager : INotifyTimeLimit, IWorldLoaded, ITick
+	{
+		readonly TimeLimitManagerInfo info;
+		LabelWidget countdownLabel;
+		public int TimeLimit;
+		public string Notification;
+
+		public TimeLimitManager(Actor self, TimeLimitManagerInfo info)
+		{
+			this.info = info;
+			Notification = info.Notification;
+
+			var tl = self.World.LobbyInfo.GlobalSettings.OptionOrDefault("timelimit", info.TimeLimitDefault.ToString());
+			if (!int.TryParse(tl, out TimeLimit))
+				TimeLimit = info.TimeLimitDefault;
+
+			// Convert from minutes to ticks
+			TimeLimit *= 60 * (1000 / self.World.Timestep);
+		}
+
+		void IWorldLoaded.WorldLoaded(World w, OpenRA.Graphics.WorldRenderer wr)
+		{
+			if (!string.IsNullOrWhiteSpace(info.CountdownLabel) && !string.IsNullOrWhiteSpace(info.CountdownText))
+			{
+				countdownLabel = Ui.Root.GetOrNull<LabelWidget>(info.CountdownLabel);
+			}
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (TimeLimit <= 0)
+				return;
+
+			var ticksPerSecond = 1000 / self.World.Timestep;
+			var ticksRemaining = TimeLimit - self.World.WorldTick;
+
+			if (countdownLabel != null && ticksRemaining > 0 && ticksRemaining % 25 == 0)
+				countdownLabel.GetText = () => info.CountdownText.F(WidgetUtils.FormatTime(ticksRemaining, true, self.World.Timestep));
+
+			if (ticksRemaining == 0)
+			{
+				foreach (var ntl in self.TraitsImplementing<INotifyTimeLimit>())
+					ntl.NotifyTimerExpired(self);
+
+				foreach (var p in self.World.Players)
+					foreach (var ntl in p.PlayerActor.TraitsImplementing<INotifyTimeLimit>())
+						ntl.NotifyTimerExpired(p.PlayerActor);
+
+				return;
+			}
+
+			if (info.SkipTimeRemainingNotifications)
+				return;
+
+			if (ticksRemaining > 0 && info.TimeLimitWarnings.Keys.All(m => ticksRemaining != m * 60 * ticksPerSecond))
+				return;
+
+			foreach (var m in info.TimeLimitWarnings.Keys)
+			{
+				if (ticksRemaining == m * 60 * ticksPerSecond)
+				{
+					Game.AddChatLine(Color.White, "Battlefield Control", Notification.F(m, m > 1 ? "s" : null));
+
+					var rp = self.World.RenderPlayer;
+					if (rp != null)
```

Comment:
This is what prevents the notification playing for observers.

This can be fixed by doing something like:
```csharp
var faction = self.World.LocalPlayer == null ? null : self.World.LocalPlayer.Faction.InternalName;
Game.Sound.PlayNotification(self.World.Map.Rules, self.World.LocalPlayer, "Speech", info.TimeLimitWarnings[m], faction);
```
(copied from the "battle control terminated" audio in the ingame menu logic).